build.gradle

```diff
@@ -1,15 +1,50 @@
-import java.text.SimpleDateFormat
-
 plugins {
     id 'fabric-loom' version '1.2-SNAPSHOT'
     id 'maven-publish'
-    id "org.ajoberstar.grgit" version "4.1.0"
+    id "org.ajoberstar.grgit" version "5.2.0"
     id 'org.jetbrains.kotlin.jvm' version '1.8.10'
     id 'org.jetbrains.kotlin.plugin.serialization' version '1.8.10'
     id 'me.fallenbreath.yamlang' version '1.3.0'
 }
 
-version = project.mod_version + "+" + new SimpleDateFormat("MMddHHmm").format(new Date())
+ext {
+    Map<String, String> environmentMap = System.getenv()
+
+    getVersionType = {
+        switch (environmentMap.REDEN_BUILD_TYPE) {
+            case "RELEASE":
+                return "stable"
+            case "BETA":
+                return "beta"
+            default:
+                return "dev"
+        }
+    }
+
+    getVersionPatch = { List paths ->
+        if (grgit == null) {
+            return 0
+        }
+
+        List latestCommits = paths.isEmpty() ? grgit.log() : grgit.log(paths: paths)
+        return latestCommits.size()
+    }
+
+    getVersionGit = { List paths ->
+        if (grgit == null) {
+            return "nogit"
+        }
+
+        List latestCommits = paths.isEmpty() ? grgit.log(maxCommits: 1) : grgit.log(paths: paths, maxCommits: 1)
+        return latestCommits.isEmpty() ? "uncommited" : "${latestCommits.get(0).id.substring(0, 7)}"
```

Comment:
why substring 0, 7?

Reply:
first to seventh char of the commit SHA
Index 0 inclusive and index 7 exclusive

Reply:
but why 7 hex digits? I prefer 8 as 4 bytes.

Reply:
> but why 7 hex digits? I prefer 8 as 4 bytes.

Because it is exactly what GitHub shows in the commit history

Reply:
well, ok

Reply:
And if you try:

> git rev-parse --short HEAD

By default, it will gives you a 7 character long SHA

> ad7e9c4